src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -2252,20 +2320,47 @@ private static async Task ExecuteResultWriteResponse(IResult? result, HttpContex
         await EnsureRequestResultNotNull(result).ExecuteAsync(httpContext);
     }
 
-    private static Task WriteJsonResponse(HttpResponse response, object? value, JsonSerializerOptions? options)
+    // This method will not check for polymorphism and will
+    // leverage the STJ polymorphism support.
+    private static Task WriteJsonResponseFast<T>(HttpResponse response, T value, JsonTypeInfo jsonTypeInfo)
+        => HttpResponseJsonExtensions.WriteAsJsonAsync(response, value, (JsonTypeInfo<T>)jsonTypeInfo, default);
+
+    private static Task WriteJsonResponse<T>(HttpResponse response, T? value, JsonSerializerOptions? options, JsonTypeInfo? jsonTypeInfo)
     {
+        var runtimeType = value is null ? typeof(object) : value.GetType();
+
+        if (jsonTypeInfo is not null && jsonTypeInfo.Type == runtimeType)
+        {
+            // In this case the polymorphism is not
+            // relevant for us and will be handled by STJ, if needed.
+            return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value!, (JsonTypeInfo<T>)jsonTypeInfo, default);
+        }
+
+        // We cannot use JsonTypeInfo here yet, waiting for https://github.com/dotnet/runtime/issues/77051
+        // What should happens if options or type info is null?
+        // var runtimeTypeInfo = GetJsonTypeInfo(options, runtimeType);
+        // return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value!, runtimeTypeInfo, default);
+
         // Call WriteAsJsonAsync() with the runtime type to serialize the runtime type rather than the declared type
         // and avoid source generators issues.
         // https://github.com/dotnet/aspnetcore/issues/43894
         // https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-polymorphism
-        return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value, value is null ? typeof(object) : value.GetType(), options, default);
+        return HttpResponseJsonExtensions.WriteAsJsonAsync(response, value, runtimeType, options, default);
     }
 
-    // Only for use with structs, use WriteJsonResponse for classes to preserve polymorphism
-    private static Task WriteJsonResponseOfT<T>(HttpResponse response, T value, JsonSerializerOptions? options)
+    private static JsonTypeInfo? GetJsonTypeInfo(JsonSerializerOptions? jsonSerializerOptions, Type type)
```

Comment:
I don't think it is reasonable to support null `JsonSerializerOptions` or null `TypeInfoResolver` at this level. It will always result in serialization so it seems like something that should fail in the configuration layer?
```suggestion
    private static JsonTypeInfo GetJsonTypeInfo(JsonSerializerOptions jsonSerializerOptions, Type type)
    {
         Debug.Assert(jsonSerializerOptions.TypeInfoResolver != null);
         Debug.Assert(jsonSerializerOptions.IsReadOnly);
         return jsonSerializerOptions.GetTypeInfo(type);
    }
```